### ISSUE TITLE
Socket.io + SSR webserver

### ISSUE THREAD

**Issue by DrakoPOD** [quasarframework/quasar#10567]
I'm trying to use Socket.io in a SSR app, I already setup it in a middleware, the problem is the socket server doesn't use the same port as the main app. I read all the docs  and ask to the community, but apparently, there is no way that I can get the server instance, only is available the app instance, but I can't call `app.listen` by my way, this will broken the entire app.

I can run socket.io in another port, but I need configure the CORS policy.

**Comment by hawkeye64** [comment]
If you want to add CORS, just write a middlewares file:
file: cors.js
```js
import cors from 'cors'
import { ssrMiddleware } from 'quasar/wrappers'

export default ssrMiddleware(({ app }) => {
  app.use(
    cors()
  )
})
```

Alternatively, you can add your own CORS:
file: cors.js
```js
import { ssrMiddleware } from 'quasar/wrappers'

export default ssrMiddleware(({ app }) => {
  app.use(function (req, res, next) {
    res.header('Access-Control-Allow-Origin', req.get('Origin') || '*')
    res.header('Access-Control-Allow-Credentials', 'true')
    res.header('Access-Control-Allow-Methods', 'GET,HEAD,PUT,PATCH,POST,DELETE')
    res.header('Access-Control-Expose-Headers', 'Content-Length')
    res.header('Access-Control-Allow-Headers', 'Accept, Authorization, Content-Type, X-Requested-With, Range')
    // res.header('X-XSS-Protection', '0')
    return next()
  })
})
```
Don't forget to add cors.js to middlewares in quasar.conf.js

https://quasar.dev/quasar-cli/developing-ssr/ssr-middleware#introduction

**Comment by hawkeye64** [comment]
For the http server instance, this comes after the middlewares are initialized. Quasar would need a post-middlewares and pass the server instance there.

**Comment by DrakoPOD** [comment]
@hawkeye64
Is possible add a post-middleware?

**Comment by hawkeye64** [comment]
It needs some discussion first with the team.

**Comment by rstoenescu** [comment]
Please create a bare minimum repo with your needs (and instructions) so we can take a look and help you.

**Comment by DrakoPOD** [comment]
@rstoenescu  Socket.io can use the same port that the host page, but I need the server instance of quasar in ssr mode. Doesn't need a repo for that:
```js
import express from "express";
import { createServer } from "http";
import { Server } from "socket.io";

const app = express();
const httpServer = createServer(app); //the server instance of express
const io = new Server(httpServer, { /* options */ });

io.on("connection", (socket) => {
  // ...
});

httpServer.listen(3000);
```

**Comment by rstoenescu** [comment]
Will be available through "@quasar/app" v3.2.0.
Haven't written the docs yet (todo tomorrow), but will tag this ticket when I do.

There's a caveat though. It will be possible ONLY for production and NOT for dev builds, due to how we are forced to use webpack-dev-server. HOWEVER, this week I'm starting to work on the new Quasar CLI with Vite which will lift a LOT of the current restrictions (but honestly not sure if you will be able to wait for it).

**Comment by rstoenescu** [comment]
Long story short, with "@quasar/app" v3.2.0 create a file named "/src-ssr/production-export.js" and inject this into it:

```js
import { createServer } from 'http'
import { Server } from 'socket.io'

export default ({ app, port, isReady }) => {
  return isReady().then(() => {
    const httpServer = createServer(app)
    const io = new Server(httpServer, { /* options */ })

    io.on("connection", (socket) => {
      // ...
    })

    httpServer.listen(port)
  })
}
```
